keras_nlp/tokenizers/sentence_piece_tokenizer.py

```diff
@@ -0,0 +1,157 @@
+        sentence_iterator=ds.as_numpy_iterator(),
+        model_writer=model,
+        vocab_size=20,
+    )
+
+    # Tokenize inputs
+    tokenizer = SentencePieceTokenizer(model_bytes=model.getvalue())
+    ds = ds.map(tokenizer)
+    ```
+    """
+
+    def __init__(
+        self,
+        model_file: str = None,
+        model_bytes: bytes = None,
+        sequence_length: int = None,
+        **kwargs,
+    ) -> None:
+        # Check dtype and provide a default.
+        if "dtype" not in kwargs or kwargs["dtype"] is None:
+            kwargs["dtype"] = tf.int32
+        else:
+            dtype = tf.dtypes.as_dtype(kwargs["dtype"])
+            if not dtype.is_integer and dtype != tf.string:
+                raise ValueError(
+                    "Output dtype must be an integer type or a string. "
+                    f"Received: dtype={dtype}"
+                )
+
+        super().__init__(**kwargs)
+
+        if model_file is None and model_bytes is None:
+            raise ValueError(
+                "One of `model_file` or `model_bytes` must be set. "
+                "Received: `model_file=None`, `model_bytes=None`."
+            )
+
+        if model_file is not None:
+            model_bytes = tf.io.gfile.GFile(model_file, "rb").read()
+        if isinstance(model_bytes, str):
+            model_bytes = base64.b64decode(model_bytes)
+
+        # Keras cannot serialize a bytestring, so we base64 encode the model
+        # byte array for saving.
+        self.model_bytes = base64.b64encode(model_bytes).decode("ascii")
+        self.sequence_length = sequence_length
+
+        self._sentence_piece = tf_text.SentencepieceTokenizer(
+            model=model_bytes,
+            out_type=self.compute_dtype,
+        )
+
+    def vocabulary_size(self) -> int:
+        """Get the size of the tokenizer vocabulary."""
+        return self._sentence_piece.vocab_size()
+
+    def id_to_token(self, id: int) -> str:
+        """Convert an integer id to a string token."""
+        return self._sentence_piece.id_to_string(id)
+
+    def token_to_id(self, token: str) -> int:
+        """Convert a string token to an integer id."""
+        return self._sentence_piece.string_to_id(token)
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                # Ideally the model would be saved as a file asset in
+                # the saved model. We have no good way to support this
+                # currently, so we save the model string in the config.
+                "model_bytes": self.model_bytes,
+                "sequence_length": self.sequence_length,
+            }
+        )
+        return config
+
+    def tokenize(self, inputs):
+        # Check if Input is Scalar or Not
+        if not isinstance(inputs, (tf.Tensor, tf.RaggedTensor)):
+            inputs = tf.convert_to_tensor(inputs)
+        scalar_input = tf.convert_to_tensor(inputs).shape.rank == 0
```

Comment:
no need to call `convert_to_tensor`? Looks it's guaranteed to be a tensor after the previous check.

Reply:
done!